pkg/orchestrator/scheduler/batch_job.go

```diff
@@ -0,0 +1,202 @@
+			evaluation.JobID, evaluation, err)
+	}
+	nonTerminalExecutions := jobState.NonTerminalExecutions()
+
+	// early exit if the job is stopped
+	if jobState.State.IsTerminal() {
+		for _, execution := range nonTerminalExecutions {
+			plan.AppendStoppedExecution(execution, model.ExecutionStateCanceled, execNotNeeded)
+		}
+		return b.planner.Process(ctx, plan)
+	}
+
+	// Retrieve the info for all the nodes that have executions for this job
+	nodeInfos, err := b.existingNodeInfos(ctx, nonTerminalExecutions)
+	if err != nil {
+		return err
+	}
+
+	// Mark executions that are running on nodes that are not healthy as failed
+	err = b.failUnhealthyExecs(ctx, nonTerminalExecutions, nodeInfos, plan)
+	if err != nil {
+		return err
+	}
+
+	job, err := b.jobStore.GetJob(ctx, evaluation.JobID)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve job %s: %w", evaluation.JobID, err)
+	}
+
+	// create new executions if needed
+	activeExecutionsCount := len(nonTerminalExecutions) - plan.StoppedExecutionsCount()
+	desiredExecutionCount := job.Spec.Deal.Concurrency - jobState.CompletedCount() - activeExecutionsCount
+	if desiredExecutionCount > 0 {
+		selectedNodes, err := b.selectNodes(ctx, &job, desiredExecutionCount)
+		if err != nil {
+			return err
+		}
+		for i := 0; i < desiredExecutionCount; i++ {
+			executionID := model.ExecutionID{
+				JobID:       job.Metadata.ID,
+				NodeID:      selectedNodes[i].PeerInfo.ID.String(),
+				ExecutionID: "e-" + uuid.NewString(),
+			}
+			execution := &model.ExecutionState{
+				JobID:            executionID.JobID,
+				NodeID:           executionID.NodeID,
+				ComputeReference: executionID.ExecutionID,
+				State:            model.ExecutionStateAskForBid,
+			}
+			plan.AppendExecution(execution)
+		}
+	}
+	// stop executions if we over-subscribed and exceeded the desired number of executions
+	if desiredExecutionCount < 0 {
+		candidateExecutions := lo.Filter(nonTerminalExecutions, func(execution *model.ExecutionState, _ int) bool {
+			return !plan.IsExecutionStopped(execution)
+		})
+		// TODO: keep track of execution ranks and kill the worst ones
+		// Using version as indicator of which execution has made more progress
+		sort.Slice(candidateExecutions, func(i, j int) bool {
+			return candidateExecutions[i].Version < candidateExecutions[j].Version
+		})
+		for i := 0; i < math.Abs(desiredExecutionCount); i++ {
+			plan.AppendStoppedExecution(candidateExecutions[i], model.ExecutionStateCanceled, execNotNeeded)
+		}
+	}
+
+	return b.planner.Process(ctx, plan)
+}
+
+// failUnhealthyExecs marks executions that are running on nodes that are not healthy as failed
+func (b *BatchJobScheduler) failUnhealthyExecs(
+	ctx context.Context, existingExecutions []*model.ExecutionState, nodeInfos map[string]*model.NodeInfo, plan *models.Plan) error {
+	for _, execution := range existingExecutions {
+		if _, ok := nodeInfos[execution.NodeID]; !ok {
```

Comment:
For this case, where the node has dropped off the network, we're marking the execution as stopped even though the compute node may not find out about it and we won't be able to connect to the compute node.  When the compute node restarts/rejoins, it'll have this execution in store.  For batch jobs like this, when the compute node restarts, I was considering marking anything active as stopped/cancelled on a reboot and trying to report the new state back to the requester node. Does it make sense to do that?